docs/developers_notes/basis_module.md

```diff
@@ -0,0 +1,130 @@
+# The Basis Module
+
+The `neurostatslib.basis` module provides objects that allow users to construct and evaluate basis functions of various types. The classes are hierarchically organized as follows:
+
+```
+Abstract Class Basis
+|
+├─ Abstract Subclass AdditiveBasis
+│
+├─ Abstract Subclass MultiplicativeBasis
+│
+├─ Abstract Subclass SplineBasis
+│   │
+│   └─ Subclass MSplineBasis
+│
+├─ Abstract Subclass RaisedCosineBasis
+│   │
+│   ├─ Subclass RaisedCosineBasisLinear
+│   │
+│   └─ Subclass RaisedCosineBasisLog
+│
+└─ Class OrthExponentialBasis
+```
+
+We've used abstract classes to ensure that any basis object inheriting from the superclass will implement the abstract methods of the superclass. This guarantees that if the inputs and outputs of those methods conform to the requirements specified by the abstract superclass, any new basis class implementation will integrate seamlessly with the basis objects already implemented in the library.
+
+The user only needs to instantiate the non-abstract subclasses located at the bottom of the hierarchy. These classes provide two public methods, `evaluate` and `evaluate_on_grid`, both of which are defined in the superclass `Basis`. These methods perform checks on both the input provided by the user and the output of the evaluation to ensure correctness, and are thus considered "safe."
+
+Additionally, a user can combine multiple objects that are subclasses of Basis, creating a higher-dimensional basis set object. Basis objects can be combined using the `__add__` and `__mul__` methods, which allow for the following syntax:
+
+```
+basis_add = basis_a + basis_b # calls basis_a.__add__(basis_b)
+
+basis_add = basis_a * basis_b # calls basis_b.__mul_(basis_b)
+```
+
+## The Class `neurostatslib.basis.Basis`
+
+### The Public Method `evaluate`
+
+The `evaluate` method checks input consistency and evaluates the basis function at some sample points. It accepts one or more numpy arrays as input, which represent the sample points at which the basis will be evaluated, and performs the following steps:
+
+1. Checks that the inputs all have the same sample size `N`, and raises a `ValueError` if this is not the case.
+2. Checks that the number of inputs matches what the basis being evaluated expects (e.g., one input for a 1-D basis, N inputs for an N-D basis, or the sum of N 1-D bases), and raises a `ValueError` if this is not the case.
+3. Checks that the output of the method does not exceed `gb_limit` GB of memory size. This size limit is specified at class initialization.
+4. Calls the `_evaluate` method on the input, which is the subclass-specific implementation of the basis set evaluation.
+5. Returns a numpy array of dimension `N x number of bases`, with each basis element evaluated at the samples.
```

Comment:
```suggestion
5. Returns a numpy array of shape `(N, n_basis_funcs)`, with each basis element evaluated at the samples.
```

two changes: use numpy shape convention (clearer than x), and change to `n_basis_funcs`, which is an argument we accept. 

Feels a bit weird to mix "mathy" (`N`) and "computer-sciencey" (`n_basis_funcs`) conventions, I'd vote for being consistently computer-sciencey, except in equations.